src/coreclr/jit/valuenum.cpp

```diff
@@ -8319,20 +8319,26 @@ void Compiler::fgValueNumberBlockAssignment(GenTree* tree)
         // SSA names in which to store VN's on defs.  We'll yield unique VN's when we read from them.
         if (lclDefSsaNum != SsaConfig::RESERVED_SSA_NUM)
         {
-            unsigned lhsLclSize = lvaLclExactSize(lhsLclNum);
-            unsigned storeSize  = lhs->GetLayout(this)->GetSize();
+            ClassLayout* const layout    = lhs->GetLayout(this);
+            unsigned           storeSize = layout->GetSize();
 
             ValueNumPair rhsVNPair = ValueNumPair();
             if (tree->OperIsInitBlkOp())
             {
-                ValueNum initObjVN = ValueNumStore::NoVN;
-                if (isEntire && rhs->IsIntegralConst(0))
+                ValueNum   initObjVN  = ValueNumStore::NoVN;
+                bool const isZeroInit = rhs->IsIntegralConst(0);
+                if (isEntire && isZeroInit)
                 {
                     // Note that it is possible to see pretty much any kind of type for the local
                     // (not just TYP_STRUCT) here because of the ASG(BLK(ADDR(LCL_VAR/FLD)), 0) form.
                     initObjVN = (lhsVarDsc->TypeGet() == TYP_STRUCT) ? vnStore->VNForZeroObj(lhsVarDsc->GetStructHnd())
                                                                      : vnStore->VNZeroForType(lhsVarDsc->TypeGet());
                 }
+                else if (isZeroInit)
+                {
+                    initObjVN = (lhs->TypeGet() == TYP_STRUCT) ? vnStore->VNForZeroObj(layout->GetClassHandle())
```

Comment:
Not all layouts will have handles. The right(-ish, it's not actually optimal, but better than the current state) fix for this is to simply replace the first argument of `VNF_ZeroObj` with a `VNForIntPtrCon(reinterpret_cast<ssize_t>(layout))`.

Reply:
I don't understand what you are suggesting here.

Reply:
Replacing `VN of CORINFO_CLASS_HANDLE` with `VN of  ClassLayout*` as the argument to `VNF_ZeroObj`.

Reply:
So, use the layout instance to classify the zeros instead of the handle, and use VNForIntPtrCon to encode the layout instance.